Sources/Commands/SwiftPackageTool.swift

```diff
@@ -134,17 +134,19 @@ extension SwiftPackageTool {
                 dryRun: dryRun
             )
 
-            if let pinsStore = swiftTool.diagnostics.wrap({ try workspace.pinsStore.load() }), let changes = changes, dryRun {
-                logPackageChanges(changes: changes, pins: pinsStore)
-            }
-
-            if !dryRun {
-                // try to load the graph which will emit any errors
+            // try to load the graph which will emit any errors
+            if !swiftTool.diagnostics.hasErrors {
```

Comment:
Technically speaking this would still emit double warnings, right?  It might not be a problem at the moment and this does seems like the safest fix that can also be put into 5.4, but this might be worth reconsidering for `main`.

Reply:
how would you restructure this? add `isEmpty` to `DiagnosticsEngine`?

Reply:
No, I mean that the code issuing the diagnostic would know whether it's already failed earlier and is just doing a retry.  Perhaps I'm misreading the code here — I'd have to look at it more closely to be able to recommend what to do, but it's the asking of the diagnostics engine that seems wrong.

Reply:
ah, this is not actually a retry. what is happening is that `workspace.updateDependencies` fails and the indication to that is that it emits a diagnostics. we could look into changing `workspace.updateDependencies` to a throwing function - which may make sense, but here I just made sure we are not calling `workspace.loadPackageGraph` if `workspace.updateDependencies` emitted errors, since they would produce addition (and in this case duplicate) errors. for additional context, calling `workspace.loadPackageGraph` here is something we added in 5.4 since `swift package update` did not show graph errors.

Reply:
I see — I had misread the code.

This looks good to me based on that understanding.